File: zig-backend/tests/abi-smoke/io_error.c
#include <lean/lean.h>

#include <stdint.h>
#include <stdio.h>

#define CHECK(cond)                                                                 \
    do {                                                                            \
        if (!(cond)) {                                                              \
            fprintf(stderr, "FAIL:%s:%d: %s\n", __FILE__, __LINE__, #cond);         \
            return 1;                                                               \
        }                                                                           \
    } while (0)

static int check_option_some(lean_object *opt, lean_object *expected) {
    lean_ctor_object *ctor = (lean_ctor_object *)opt;
    CHECK(opt != NULL);
    CHECK(!lean_is_scalar(opt));
    CHECK(lean_obj_tag(opt) == 1);
    CHECK(ctor->m_header.m_other == 1);
    CHECK(ctor->m_objs[0] == expected);
    return 0;
}

static int check_heap_ctor(lean_object *result, unsigned tag, unsigned num_objs) {
    lean_ctor_object *ctor = (lean_ctor_object *)result;
    CHECK(result != NULL);
    CHECK(!lean_is_scalar(result));
    CHECK(lean_obj_tag(result) == tag);
    CHECK(ctor->m_header.m_other == num_objs);
    return 0;
}

static uint32_t get_ctor_uint32(lean_object *result, unsigned num_objs) {
    lean_ctor_object *ctor = (lean_ctor_object *)result;
    return *(uint32_t *)((uint8_t *)ctor->m_objs + ((size_t)num_objs * sizeof(void *)));
}

static int check_details_only(lean_object *result, unsigned tag, uint32_t code, lean_object *details) {
    lean_ctor_object *ctor = (lean_ctor_object *)result;
    CHECK(check_heap_ctor(result, tag, 1) == 0);
    CHECK(ctor->m_objs[0] == details);
    CHECK(get_ctor_uint32(result, 1) == code);
    return 0;
}

static int check_option_none(lean_object *result, unsigned tag, uint32_t code, lean_object *details) {
    lean_ctor_object *ctor = (lean_ctor_object *)result;
    CHECK(check_heap_ctor(result, tag, 2) == 0);
    CHECK(ctor->m_objs[0] == lean_box(0));
    CHECK(ctor->m_objs[1] == details);
    CHECK(get_ctor_uint32(result, 2) == code);
    return 0;
}

static int check_option_some_ctor(
    lean_object *result,
    unsigned tag,
    lean_object *filename,
    uint32_t code,
    lean_object *details
) {
    lean_ctor_object *ctor = (lean_ctor_object *)result;
    CHECK(check_heap_ctor(result, tag, 2) == 0);
    CHECK(check_option_some(ctor->m_objs[0], filename) == 0);
    CHECK(ctor->m_objs[1] == details);
    CHECK(get_ctor_uint32(result, 2) == code);
    return 0;
}

static int check_direct_filename(
    lean_object *result,
    unsigned tag,
    lean_object *filename,
    uint32_t code,
    lean_object *details
) {
    lean_ctor_object *ctor = (lean_ctor_object *)result;
    CHECK(check_heap_ctor(result, tag, 2) == 0);
    CHECK(ctor->m_objs[0] == filename);
    CHECK(ctor->m_objs[1] == details);
    CHECK(get_ctor_uint32(result, 2) == code);
    return 0;
}

int main(void) {
    const uint32_t code = 0xdecafbadU;

    {
        lean_object *details = lean_mk_string("already exists");
        lean_object *result = lean_mk_io_error_already_exists(code, details);
        CHECK(check_option_none(result, 0, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *filename = lean_mk_string("exists.txt");
        lean_object *details = lean_mk_string("already exists file");
        lean_object *result = lean_mk_io_error_already_exists_file(filename, code, details);
        CHECK(check_option_some_ctor(result, 0, filename, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *result = lean_mk_io_error_eof(lean_box(0));
        CHECK(result != NULL);
        CHECK(lean_is_scalar(result));
        CHECK(lean_obj_tag(result) == 17);
    }

    {
        lean_object *details = lean_mk_string("hardware fault");
        lean_object *result = lean_mk_io_error_hardware_fault(code, details);
        CHECK(check_details_only(result, 5, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("illegal operation");
        lean_object *result = lean_mk_io_error_illegal_operation(code, details);
        CHECK(check_details_only(result, 7, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("inappropriate type");
        lean_object *result = lean_mk_io_error_inappropriate_type(code, details);
        CHECK(check_option_none(result, 15, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *filename = lean_mk_string("dir");
        lean_object *details = lean_mk_string("inappropriate type file");
        lean_object *result = lean_mk_io_error_inappropriate_type_file(filename, code, details);
        CHECK(check_option_some_ctor(result, 15, filename, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *filename = lean_mk_string("interrupt");
        lean_object *details = lean_mk_string("interrupted");
        lean_object *result = lean_mk_io_error_interrupted(filename, code, details);
        CHECK(check_direct_filename(result, 10, filename, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("invalid argument");
        lean_object *result = lean_mk_io_error_invalid_argument(code, details);
        CHECK(check_option_none(result, 12, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *filename = lean_mk_string("bad.txt");
        lean_object *details = lean_mk_string("invalid argument file");
        lean_object *result = lean_mk_io_error_invalid_argument_file(filename, code, details);
        CHECK(check_option_some_ctor(result, 12, filename, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *filename = lean_mk_string("missing.txt");
        lean_object *details = lean_mk_string("no file or directory");
        lean_object *result = lean_mk_io_error_no_file_or_directory(filename, code, details);
        CHECK(check_direct_filename(result, 11, filename, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("no such thing");
        lean_object *result = lean_mk_io_error_no_such_thing(code, details);
        CHECK(check_option_none(result, 16, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *filename = lean_mk_string("ghost");
        lean_object *details = lean_mk_string("no such thing file");
        lean_object *result = lean_mk_io_error_no_such_thing_file(filename, code, details);
        CHECK(check_option_some_ctor(result, 16, filename, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("other error");
        lean_object *result = lean_mk_io_error_other_error(code, details);
        CHECK(check_details_only(result, 1, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("permission denied");
        lean_object *result = lean_mk_io_error_permission_denied(code, details);
        CHECK(check_option_none(result, 13, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *filename = lean_mk_string("secret");
        lean_object *details = lean_mk_string("permission denied file");
        lean_object *result = lean_mk_io_error_permission_denied_file(filename, code, details);
        CHECK(check_option_some_ctor(result, 13, filename, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("protocol error");
        lean_object *result = lean_mk_io_error_protocol_error(code, details);
        CHECK(check_details_only(result, 8, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("resource busy");
        lean_object *result = lean_mk_io_error_resource_busy(code, details);
        CHECK(check_details_only(result, 2, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("resource exhausted");
        lean_object *result = lean_mk_io_error_resource_exhausted(code, details);
        CHECK(check_option_none(result, 14, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *filename = lean_mk_string("full");
        lean_object *details = lean_mk_string("resource exhausted file");
        lean_object *result = lean_mk_io_error_resource_exhausted_file(filename, code, details);
        CHECK(check_option_some_ctor(result, 14, filename, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("resource vanished");
        lean_object *result = lean_mk_io_error_resource_vanished(code, details);
        CHECK(check_details_only(result, 3, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("time expired");
        lean_object *result = lean_mk_io_error_time_expired(code, details);
        CHECK(check_details_only(result, 9, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("unsatisfied constraints");
        lean_object *result = lean_mk_io_error_unsatisfied_constraints(code, details);
        CHECK(check_details_only(result, 6, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *details = lean_mk_string("unsupported operation");
        lean_object *result = lean_mk_io_error_unsupported_operation(code, details);
        CHECK(check_details_only(result, 4, code, details) == 0);
        lean_dec(result);
    }

    {
        lean_object *msg = lean_mk_string("user error");
        lean_object *result = lean_mk_io_user_error(msg);
        lean_ctor_object *ctor = (lean_ctor_object *)result;
        CHECK(check_heap_ctor(result, 18, 1) == 0);
        CHECK(ctor->m_objs[0] == msg);
        lean_dec(result);
    }

    return 0;
}

File: src/runtime/zig/io_error.zig
// Copyright (c) 2026 Lean FRO, LLC. All rights reserved.
// Released under Apache 2.0 license as described in the file LICENSE.

const std = @import("std");
const testing = std.testing;
const alloc = @import("alloc.zig");
const ctor = @import("ctor.zig");
const lean = @import("lean_object.zig");
const object = @import("object.zig");
const rc = @import("rc.zig");
const runtime_options = @import("runtime_options");

const pointer_bytes: c_uint = @sizeOf(?*anyopaque);

fn mkOptionSome(value: *anyopaque) *anyopaque {
    const some = alloc.lean_alloc_ctor(1, 1, 0);
    ctor.lean_ctor_set(some, 0, value);
    return some;
}

fn mkOptionFilenameError(tag: c_uint, filename: ?*anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    const result = alloc.lean_alloc_ctor(tag, 2, 4);
    ctor.lean_ctor_set(result, 0, if (filename) |name| mkOptionSome(name) else object.lean_box(0));
    ctor.lean_ctor_set(result, 1, details);
    ctor.lean_ctor_set_uint32(result, pointer_bytes * 2, os_code);
    return result;
}

fn mkDirectFilenameError(tag: c_uint, filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    const result = alloc.lean_alloc_ctor(tag, 2, 4);
    ctor.lean_ctor_set(result, 0, filename);
    ctor.lean_ctor_set(result, 1, details);
    ctor.lean_ctor_set_uint32(result, pointer_bytes * 2, os_code);
    return result;
}

fn mkDetailsOnlyError(tag: c_uint, os_code: u32, details: *anyopaque) *anyopaque {
    const result = alloc.lean_alloc_ctor(tag, 1, 4);
    ctor.lean_ctor_set(result, 0, details);
    ctor.lean_ctor_set_uint32(result, pointer_bytes, os_code);
    return result;
}

pub fn lean_mk_io_error_already_exists(os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(0, null, os_code, details);
}

pub fn lean_mk_io_error_already_exists_file(filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(0, filename, os_code, details);
}

pub fn lean_mk_io_error_eof(unit: *anyopaque) *anyopaque {
    _ = unit;
    return object.lean_box(17).?;
}

pub fn lean_mk_io_error_hardware_fault(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(5, os_code, details);
}

pub fn lean_mk_io_error_illegal_operation(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(7, os_code, details);
}

pub fn lean_mk_io_error_inappropriate_type(os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(15, null, os_code, details);
}

pub fn lean_mk_io_error_inappropriate_type_file(filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(15, filename, os_code, details);
}

pub fn lean_mk_io_error_interrupted(filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    return mkDirectFilenameError(10, filename, os_code, details);
}

pub fn lean_mk_io_error_invalid_argument(os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(12, null, os_code, details);
}

pub fn lean_mk_io_error_invalid_argument_file(filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(12, filename, os_code, details);
}

pub fn lean_mk_io_error_no_file_or_directory(filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    return mkDirectFilenameError(11, filename, os_code, details);
}

pub fn lean_mk_io_error_no_such_thing(os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(16, null, os_code, details);
}

pub fn lean_mk_io_error_no_such_thing_file(filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(16, filename, os_code, details);
}

pub fn lean_mk_io_error_other_error(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(1, os_code, details);
}

pub fn lean_mk_io_error_permission_denied(os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(13, null, os_code, details);
}

pub fn lean_mk_io_error_permission_denied_file(filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(13, filename, os_code, details);
}

pub fn lean_mk_io_error_protocol_error(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(8, os_code, details);
}

pub fn lean_mk_io_error_resource_busy(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(2, os_code, details);
}

pub fn lean_mk_io_error_resource_exhausted(os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(14, null, os_code, details);
}

pub fn lean_mk_io_error_resource_exhausted_file(filename: *anyopaque, os_code: u32, details: *anyopaque) *anyopaque {
    return mkOptionFilenameError(14, filename, os_code, details);
}

pub fn lean_mk_io_error_resource_vanished(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(3, os_code, details);
}

pub fn lean_mk_io_error_time_expired(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(9, os_code, details);
}

pub fn lean_mk_io_error_unsatisfied_constraints(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(6, os_code, details);
}

pub fn lean_mk_io_error_unsupported_operation(os_code: u32, details: *anyopaque) *anyopaque {
    return mkDetailsOnlyError(4, os_code, details);
}

pub fn lean_mk_io_user_error(msg: *anyopaque) *anyopaque {
    const result = alloc.lean_alloc_ctor(18, 1, 0);
    ctor.lean_ctor_set(result, 0, msg);
    return result;
}

pub export fn lean_mk_io_error_already_exists_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_already_exists(os_code, details);
}

pub export fn lean_mk_io_error_already_exists_file_zig_impl(filename: *anyopaque, os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_already_exists_file(filename, os_code, details);
}

pub export fn lean_mk_io_error_eof_zig_impl(unit: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_eof(unit);
}

pub export fn lean_mk_io_error_hardware_fault_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_hardware_fault(os_code, details);
}

pub export fn lean_mk_io_error_illegal_operation_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_illegal_operation(os_code, details);
}

pub export fn lean_mk_io_error_inappropriate_type_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_inappropriate_type(os_code, details);
}

pub export fn lean_mk_io_error_inappropriate_type_file_zig_impl(filename: *anyopaque, os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_inappropriate_type_file(filename, os_code, details);
}

pub export fn lean_mk_io_error_interrupted_zig_impl(filename: *anyopaque, os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_interrupted(filename, os_code, details);
}

pub export fn lean_mk_io_error_invalid_argument_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_invalid_argument(os_code, details);
}

pub export fn lean_mk_io_error_invalid_argument_file_zig_impl(filename: *anyopaque, os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_invalid_argument_file(filename, os_code, details);
}

pub export fn lean_mk_io_error_no_file_or_directory_zig_impl(filename: *anyopaque, os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_no_file_or_directory(filename, os_code, details);
}

pub export fn lean_mk_io_error_no_such_thing_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_no_such_thing(os_code, details);
}

pub export fn lean_mk_io_error_no_such_thing_file_zig_impl(filename: *anyopaque, os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_no_such_thing_file(filename, os_code, details);
}

pub export fn lean_mk_io_error_other_error_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_other_error(os_code, details);
}

pub export fn lean_mk_io_error_permission_denied_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_permission_denied(os_code, details);
}

pub export fn lean_mk_io_error_permission_denied_file_zig_impl(filename: *anyopaque, os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_permission_denied_file(filename, os_code, details);
}

pub export fn lean_mk_io_error_protocol_error_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_protocol_error(os_code, details);
}

pub export fn lean_mk_io_error_resource_busy_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_resource_busy(os_code, details);
}

pub export fn lean_mk_io_error_resource_exhausted_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_resource_exhausted(os_code, details);
}

pub export fn lean_mk_io_error_resource_exhausted_file_zig_impl(filename: *anyopaque, os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_resource_exhausted_file(filename, os_code, details);
}

pub export fn lean_mk_io_error_resource_vanished_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_resource_vanished(os_code, details);
}

pub export fn lean_mk_io_error_time_expired_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_time_expired(os_code, details);
}

pub export fn lean_mk_io_error_unsatisfied_constraints_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_unsatisfied_constraints(os_code, details);
}

pub export fn lean_mk_io_error_unsupported_operation_zig_impl(os_code: u32, details: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_error_unsupported_operation(os_code, details);
}

pub export fn lean_mk_io_user_error_zig_impl(msg: *anyopaque) callconv(.c) *anyopaque {
    return lean_mk_io_user_error(msg);
}
extern fn lean_mk_string_unchecked(s: [*:0]const u8, sz: usize, len: usize) callconv(.c) *anyopaque;

fn lean_io_error_to_string_impl(err: *anyopaque) callconv(.c) *anyopaque {
    const tag = object.lean_obj_tag(err);
    if (tag == 17) {
        return lean_mk_string_unchecked("end of file".ptr, 11, 11);
    }
    if (tag == 18) {
        const msg = ctor.lean_ctor_get(err, 0).?;
        rc.lean_inc(msg);
        return msg;
    }
    return lean_mk_string_unchecked("IO error".ptr, 8, 8);
}
comptime {
    if (runtime_options.export_lean_helpers) {
        @export(&lean_io_error_to_string_impl, .{ .name = "lean_io_error_to_string" });
    }
}

fn expectOptionSome(option_value: ?*anyopaque, expected: ?*anyopaque) !void {
    try testing.expect(option_value != null);
    try testing.expect(!object.lean_is_scalar(option_value));
    try testing.expectEqual(@as(c_uint, 1), object.lean_obj_tag(option_value));
    try testing.expectEqual(expected, ctor.lean_ctor_get(option_value.?, 0));
}

fn expectScalarCode(result: *anyopaque, object_count: c_uint, expected: u32) !void {
    try testing.expectEqual(expected, ctor.lean_ctor_get_uint32(result, pointer_bytes * object_count));
}

fn expectDetailsOnly(result: *anyopaque, tag: c_uint, expected_code: u32, expected_details: ?*anyopaque) !void {
    try testing.expect(!object.lean_is_scalar(result));
    try testing.expectEqual(tag, object.lean_obj_tag(result));
    try testing.expectEqual(expected_details, ctor.lean_ctor_get(result, 0));
    try expectScalarCode(result, 1, expected_code);
}

fn expectOptionFilename(result: *anyopaque, tag: c_uint, filename: ?*anyopaque, expected_code: u32, expected_details: ?*anyopaque) !void {
    try testing.expect(!object.lean_is_scalar(result));
    try testing.expectEqual(tag, object.lean_obj_tag(result));
    try expectOptionSome(ctor.lean_ctor_get(result, 0), filename);
    try testing.expectEqual(expected_details, ctor.lean_ctor_get(result, 1));
    try expectScalarCode(result, 2, expected_code);
}

fn expectOptionNone(result: *anyopaque, tag: c_uint, expected_code: u32, expected_details: ?*anyopaque) !void {
    try testing.expect(!object.lean_is_scalar(result));
    try testing.expectEqual(tag, object.lean_obj_tag(result));
    try testing.expectEqual(object.lean_box(0), ctor.lean_ctor_get(result, 0));
    try testing.expectEqual(expected_details, ctor.lean_ctor_get(result, 1));
    try expectScalarCode(result, 2, expected_code);
}

fn expectDirectFilename(result: *anyopaque, tag: c_uint, filename: ?*anyopaque, expected_code: u32, expected_details: ?*anyopaque) !void {
    try testing.expect(!object.lean_is_scalar(result));
    try testing.expectEqual(tag, object.lean_obj_tag(result));
    try testing.expectEqual(filename, ctor.lean_ctor_get(result, 0));
    try testing.expectEqual(expected_details, ctor.lean_ctor_get(result, 1));
    try expectScalarCode(result, 2, expected_code);
}

fn mkString(bytes: [:0]const u8) *anyopaque {
    const size = bytes.len + 1;
    const ptr = alloc.lean_alloc_object(@sizeOf(lean.lean_string_object) + size);
    const str: *lean.lean_string_object = @ptrCast(@alignCast(ptr));
    str.m_header = .{
        .m_rc = 1,
        .m_cs_sz = 0,
        .m_other = 0,
        .m_tag = lean.LeanString,
    };
    str.m_size = size;
    str.m_capacity = size;
    str.m_length = bytes.len;
    const data: [*]u8 = @ptrCast(&str.m_data);
    @memcpy(data[0..bytes.len], bytes[0..bytes.len]);
    data[bytes.len] = 0;
    return ptr;
}

fn decIfHeap(o: ?*anyopaque) void {
    if (o) |value| {
        if (!object.lean_is_scalar(value)) {
            rc.lean_dec(value);
        }
    }
}

test "io error constructors mirror stage1 generated layouts" {
    const os_code: u32 = 0xdecafbad;

    {
        const details = mkString("already exists");
        const result = lean_mk_io_error_already_exists(os_code, details);
        defer decIfHeap(result);
        try expectOptionNone(result, 0, os_code, details);
    }

    {
        const filename = mkString("exists.txt");
        const details = mkString("already exists file");
        const result = lean_mk_io_error_already_exists_file(filename, os_code, details);
        defer decIfHeap(result);
        try expectOptionFilename(result, 0, filename, os_code, details);
    }

    {
        const result = lean_mk_io_error_eof(object.lean_box(0).?);
        try testing.expect(object.lean_is_scalar(result));
        try testing.expectEqual(@as(c_uint, 17), object.lean_obj_tag(result));
    }

    {
        const details = mkString("hardware fault");
        const result = lean_mk_io_error_hardware_fault(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 5, os_code, details);
    }

    {
        const details = mkString("illegal operation");
        const result = lean_mk_io_error_illegal_operation(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 7, os_code, details);
    }

    {
        const details = mkString("inappropriate type");
        const result = lean_mk_io_error_inappropriate_type(os_code, details);
        defer decIfHeap(result);
        try expectOptionNone(result, 15, os_code, details);
    }

    {
        const filename = mkString("dir");
        const details = mkString("inappropriate type file");
        const result = lean_mk_io_error_inappropriate_type_file(filename, os_code, details);
        defer decIfHeap(result);
        try expectOptionFilename(result, 15, filename, os_code, details);
    }

    {
        const filename = mkString("interrupt");
        const details = mkString("interrupted");
        const result = lean_mk_io_error_interrupted(filename, os_code, details);
        defer decIfHeap(result);
        try expectDirectFilename(result, 10, filename, os_code, details);
    }

    {
        const details = mkString("invalid argument");
        const result = lean_mk_io_error_invalid_argument(os_code, details);
        defer decIfHeap(result);
        try expectOptionNone(result, 12, os_code, details);
    }

    {
        const filename = mkString("bad.txt");
        const details = mkString("invalid argument file");
        const result = lean_mk_io_error_invalid_argument_file(filename, os_code, details);
        defer decIfHeap(result);
        try expectOptionFilename(result, 12, filename, os_code, details);
    }

    {
        const filename = mkString("missing.txt");
        const details = mkString("no file or directory");
        const result = lean_mk_io_error_no_file_or_directory(filename, os_code, details);
        defer decIfHeap(result);
        try expectDirectFilename(result, 11, filename, os_code, details);
    }

    {
        const details = mkString("no such thing");
        const result = lean_mk_io_error_no_such_thing(os_code, details);
        defer decIfHeap(result);
        try expectOptionNone(result, 16, os_code, details);
    }

    {
        const filename = mkString("ghost");
        const details = mkString("no such thing file");
        const result = lean_mk_io_error_no_such_thing_file(filename, os_code, details);
        defer decIfHeap(result);
        try expectOptionFilename(result, 16, filename, os_code, details);
    }

    {
        const details = mkString("other error");
        const result = lean_mk_io_error_other_error(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 1, os_code, details);
    }

    {
        const details = mkString("permission denied");
        const result = lean_mk_io_error_permission_denied(os_code, details);
        defer decIfHeap(result);
        try expectOptionNone(result, 13, os_code, details);
    }

    {
        const filename = mkString("secret");
        const details = mkString("permission denied file");
        const result = lean_mk_io_error_permission_denied_file(filename, os_code, details);
        defer decIfHeap(result);
        try expectOptionFilename(result, 13, filename, os_code, details);
    }

    {
        const details = mkString("protocol error");
        const result = lean_mk_io_error_protocol_error(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 8, os_code, details);
    }

    {
        const details = mkString("resource busy");
        const result = lean_mk_io_error_resource_busy(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 2, os_code, details);
    }

    {
        const details = mkString("resource exhausted");
        const result = lean_mk_io_error_resource_exhausted(os_code, details);
        defer decIfHeap(result);
        try expectOptionNone(result, 14, os_code, details);
    }

    {
        const filename = mkString("full");
        const details = mkString("resource exhausted file");
        const result = lean_mk_io_error_resource_exhausted_file(filename, os_code, details);
        defer decIfHeap(result);
        try expectOptionFilename(result, 14, filename, os_code, details);
    }

    {
        const details = mkString("resource vanished");
        const result = lean_mk_io_error_resource_vanished(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 3, os_code, details);
    }

    {
        const details = mkString("time expired");
        const result = lean_mk_io_error_time_expired(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 9, os_code, details);
    }

    {
        const details = mkString("unsatisfied constraints");
        const result = lean_mk_io_error_unsatisfied_constraints(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 6, os_code, details);
    }

    {
        const details = mkString("unsupported operation");
        const result = lean_mk_io_error_unsupported_operation(os_code, details);
        defer decIfHeap(result);
        try expectDetailsOnly(result, 4, os_code, details);
    }

    {
        const msg = mkString("user error");
        const result = lean_mk_io_user_error(msg);
        defer decIfHeap(result);
        try testing.expect(!object.lean_is_scalar(result));
        try testing.expectEqual(@as(c_uint, 18), object.lean_obj_tag(result));
        try testing.expectEqual(msg, ctor.lean_ctor_get(result, 0));
    }
}
